fix: make ntl dev wait for local server to be ready

eduardoboucas opened the pull request:
**- Summary**

When running `ntl dev`, we wait for the port of the local server to open before launching the browser session. However, there is sometimes a delay between the port being open and the server actually being ready to receive requests. This means that the page we open may be blank and unable to recover without a refresh.

I'm able to reproduce this consistently with a fresh Gatsby site.

The solution proposed in this PR is to poll the local server with `HEAD` requests. Once one is completed successfully, we consider the server to be ready and only then launch the browser session.

Closes #404

**- Description for the changelog**

fix: make ntl dev wait for local server to be ready

**- A picture of a cute animal (not mandatory but encouraged)**

![samoyed-outdoors-gallery-8-min](https://user-images.githubusercontent.com/4162329/105373917-11e5e800-5bff-11eb-8b53-60849d74a8ba.jpg)

erezrokah commented on src/commands/dev/index.js:
```diff
@@ -82,7 +91,30 @@ const startFrameworkServer = async function ({ settings, log, exit }) {
+      return true
+    }
+
+    await waitFor(waitForServerToRespond, {
```
Can we skip this `waitFor` if `open` is `false`?

eduardoboucas replied:
Absolutely.

erezrokah commented on src/commands/dev/index.js:
```diff
@@ -82,7 +91,30 @@ const startFrameworkServer = async function ({ settings, log, exit }) {
+      try {
+        await fetch(`http://localhost:${settings.frameworkPort}`, {
+          method: 'HEAD',
+          timeout: SERVER_POLL_REQUEST_TIMEOUT,
```
I'm not sure the `fetch` API allows passing a timeout.

We can consider using https://github.com/sindresorhus/got but that's out of scope for this PR.

eduardoboucas replied:
> Hi @eduardoboucas - I like the approach in this PR!
> I have 2 questions/suggestions:
> 
> 1. Can we verify a few other framework servers accept `HEAD` requests?
> 2. If `waitFor` times out and `open` is `true`, how about we print a warning instead of failing?

Both great suggestions, thanks @erezrokah!

1. I tested a few, happy to test a few more. Would you feel more comfortable if we used a `GET` request instead?
2. I was actually experimenting with this locally. Since you brought it up, we should definitely do it.

eduardoboucas replied:
I think it exists in the version of `node-fetch` we're using, but it was [removed in version 3](https://github.com/node-fetch/node-fetch/blob/cbd4d895767e33491219e40013dd8daa5c7ac024/docs/v3-UPGRADE-GUIDE.md#the-timeout-option-was-removed). Definitely something we need to take into account if/when upgrading.

erezrokah replied:
Oh that's very helpful to know thanks!

erezrokah reviewed (changes requested):
Hi @eduardoboucas - I like the approach in this PR!
I have 2 questions/suggestions:
1. Can we verify a few other framework servers accept `HEAD` requests?
2. If `waitFor` times out and `open` is `true`, how about we print a warning instead of failing?

eduardoboucas commented:
> Hi @eduardoboucas - I like the approach in this PR!
> I have 2 questions/suggestions:
> 
> 1. Can we verify a few other framework servers accept `HEAD` requests?
> 2. If `waitFor` times out and `open` is `true`, how about we print a warning instead of failing?

Both great suggestions, thanks @erezrokah!

1. I tested a few, happy to test a few more. Would you feel more comfortable if we used a `GET` request instead?
2. I was actually experimenting with this locally. Since you brought it up, we should definitely do it.

eduardoboucas commented:
I've addressed your suggestions in https://github.com/netlify/cli/pull/1795/commits/f85789365390c40ffe285d0aed5389c656eb9764. I also reduced the overall timeout significantly, so that we don't wait for too long if the server doesn't respond to the requests.

erezrokah commented:
> I've addressed your suggestions in [f857893](https://github.com/netlify/cli/commit/f85789365390c40ffe285d0aed5389c656eb9764). I also reduced the overall timeout significantly, so that we don't wait for too long if the server doesn't respond to the requests.

This looks great!

erezrokah commented:
Also closes https://github.com/netlify/cli/issues/1703
